Surface campaign name over ID

erxclau opened the pull request:
#### What's this PR do?

#### Why are we doing this? How does it help us?

#### How should this be manually tested?

#### How should this change be communicated to end users?

#### Are there any smells or added technical debt to note?

#### What are the relevant tickets?

#### Have you done the following, if applicable:
***(optional: add explanation between parentheses)***

* [ ] Added automated tests? *( )*
* [ ] Tested manually on mobile? *( )*
* [ ] Checked BrowserStack? *( )*
* [ ] Checked for performance implications? *( )*
* [ ] Checked accessibility? *( )*
* [ ] Checked for security implications? *( )*
* [ ] Updated the documentation/wiki? *( )*

#### TODOs / next steps:

* [ ] *your TODO here*

ashley-hebler reviewed (commented):
@erxclau 
Thank you for fixing that pre-commit issue; I'm sure others would eventually run into that. 

As for the new description value we're sending, I think that might actually be a stripe property. We can offline more about this, but I realized we can see our test donations in a salesforce sandbox and after testing with some test campaign ids I found there, I can confirm that's not what's printing in Slack. This is a little more in-the-weeds of Salesforce than I originally understood so I consulted with @x110dc via our wiki doc to check if this is even possible and that I'm thinking about this the right way.

P.S. If it is possible I LOVE that you've captured all the scenarios in the test coverage. That'll be great to have.

erxclau commented:
There have been significant changes since the OP so here are some more details about what's changed. 

#### What's this PR do?

As a general overview, this PR adds some more validation to an opportunity's campaign id and queries Salesforce for the campaign name when we want to notify Slack.

- Bumps `black` version that we use for pre-commit linting.
   - See OP. This changes `.pre-commit-config.yaml` and `dev-requirements.txt`.
- Creates a `CampaignMixin` class that applies to `Opportunity` and `RDO`.
   - This mixin has two functions: `has_invalid_campaign_id_format` and `get_campaign_name`.
   - `has_invalid_campaign_id_format` is a regex validator that returns whether or not the opportunity's campaign ID follows the expected format. According to Daniel, all campaign IDs are either 15 or 18 characters and are all alphanumerical. I chose to use regex because it made things more succinct but we could have also used the built-in `len` and `isalphanum` functions. 
      - Note that I'm passing in `str(self.campaign_id)` as the second argument. This is because the regex match function expects a string as the second argument. If `campaign.id` is None, an error would be thrown without the string cast. (None as a string is "None" so it does not pass the validator based on the length constraint).
   - `get_campaign_name` is a query function that fetches the campaign name from Salesforce. [Salesforce documentation](https://developer.salesforce.com/docs/atlas.en-us.object_reference.meta/object_reference/sforce_api_objects_campaign.htm) says that a Campaign name is required. Given that, I think we can assume that if the query successfully returns, a proper campaign name will be given to us. 
       - I handle an exception here for if an invalid ID is passed to the query. I'll talk about this more in a bit, but I don't think this exception will ever be reached because of some other error handling we do. But, in the case that we do reach this point, we just set the campaign id to None and return None as well.
- Adds more error handling to `Opportunity.save` and `RDO.save`
   - The `save` function essentially commits an opportunity to Salesforce. 
   - One minor change I did was move where we do `self.name = self.name[:80]`. There was a previous note that said "TODO this will fail if name hasn't been set." I moved the line to be underneath a name check so we don't have to worry about name not being set.
   - I call the `has_invalid_campaign_id_format` method before we try to save to Salesforce. If the ID is invalid, then we set it to None.
   - When we try to save, there is already an exception that handles if the campaign ID is invalid. Note that it is possible to reach this exception, even with the regex validator (i.e., ID that is alphanumeric and has the correct length but does not actually correspond to an existing campaign object in the Salesforce database).
   - I add a new exception handler here for `INSUFFICIENT_ACCESS_ON_CROSS_REFERENCE_ENTITY`. This is the error we ran into in #902. It seems like this error is thrown when we are trying to link an opportunity to a campaign, but the ID actually points to an object of a different type ([see this post for other cases](https://metillium.com/2016/03/invalid-cross-reference-id-error/)). I'm handling this by setting the campaign id to None and retrying the save.
- Retrieves the campaign name in `notify_slack`
  - If the campaign id for an opportunity exists, set `self.campaign_name` to the return value of `get_campaign_name`
  - Note that the `save` function is always called before we invoke the `notify_slack` function. This means that we have two layers of campaign id validation (our regex validator and the internal Salesforce validators that throw errors for invalid ids) before we try to query for a campaign name. Thus, I think it's fairly unlikely that the exception we handle in `get_campaign_name` will ever get reached but I guess it doesn't hurt to have it.
  - Retrieving the campaign name here is the most appropriate because we're only using the name in the Slack context.
- Edits `construct_slack_message`
   - I did some refactoring here. Previously we had constructs like this:
     ```python
     getattr(rdo, "encouraged_by", False) or getattr(opportunity, "encouraged_by", False) or ""
     ```
      I simplified this by creating a new variable `opp` that is set to `opportunity or rdo` such that we can do this:
      ```python
      getattr(opp, "encouraged_by", False) or ""
      ```
   - I also reordered the variables here to match the order in which they appear in the message. 
   - Lastly, I replace the campaign id with the campaign name. The OP had a fallback case where we would use the id instead of the name if the name did not exist. I no longer do this fallback because Salesforce tells us that the campaign name is required. So, if the name does not exist, that probably means there is no campaign associated with the opportunity.
- Adds new testing for `has_invalid_campaign_id_format` and `construct_slack_message`.
  - This is pretty straightforward. Note that with no longer having the fallback, I only have test cases for if there is or is not a campaign name.

#### How should this be manually tested?

There are a few cases to test:
- Valid Salesforce campaign
  - The campaign in our Salesforce sandbox has an id of `70117000001umA6AAI`. Use this as the `campaignId` query parameter.
  - Note that `70117000001umA6` (the same id with the last three characters missing) also seems to work and refers to the same campaign. I guess 15 and 18 character ids are related to each other. 
  - This will produce a successful donation and will send a Slack message with the campaign name "Test Campaign"
- Invalid regex validation
   - Try a `campaignId` parameter with an invalid length (!= 15 and != 18) or invalid characters (non-alphanumeric).
   - This will produce a successful donation and will send a Slack message with no campaign name.
- Valid regex validation but invalid Salesforce campaign
   - Use an alphanumeric id with 15 or 18 characters that is not our test campaign id.
   - This will produce a successful donation and will send a Slack message with no campaign name.
- `INSUFFICIENT_ACCESS_ON_CROSS_REFERENCE_ENTITY`
   - We can force this by using `70117000001umA7` or `701460000007V3Q` as the `campaignId` query parameter.
   - This will produce a successful donation and will send a Slack message with no campaign name.
- No campaign id query parameter
   - This will produce a successful donation and will send a Slack message with no campaign name.

#### Are there any smells or added technical debt to note?

No

#### What are the relevant tickets?

<https://airtable.com/appyo1zuQd8f4hBVx/tbloNZu8GkM52NKFR/viwPxWENGdckjxRrP/recG82vrj2bshV46r?blocks=hide>

#### Have you done the following, if applicable:
***(optional: add explanation between parentheses)***

* [x] Added automated tests? *(See above)*
* [x] Checked for performance implications? *(This doesn't affect performance for the end user because we do SO/Slack stuff in the background.)*
* [x] Checked for security implications? *(The regex validator is pretty airtight and Salesforce also checks the validity of the campaign ID before we ever query for the campaign name)*

djpeacher reviewed (approved):
LGMT! I did make one suggestion above, but looks good otherwise! Tests pass and slack logging works! 🎉 

> I do have to duplicate some code in Opportunity and RDO. There are a lot of commonalities between the two classes so it may be worthwhile to make some sort of super class for them.

Yes, I completely agree! There seems to be a lot of duplication here that could be put into a parent class.

erxclau commented on npsp.py:
```diff
@@ -31,6 +32,30 @@
+            match("^([a-zA-Z0-9]{15}|[a-zA-Z0-9]{18})$", str(self.campaign_id)) is None
+        )
+
+    def get_campaign_name(self):
+        query = f"SELECT Name FROM Campaign WHERE Id = '{self.campaign_id}'"
+        try:
+            response = self.sf.query(query)
+        except SalesforceException as e:
+            if e.content["errorCode"] == "INVALID_QUERY_FILTER_OPERATOR":
+                logging.warning(
+                    f"could not get campaign name with campaign ID; continuing..."
+                )
+                self.campaign_id = None
+                return None
+            else:
+                logging.error(e.response.text)
+                raise
+        else:
+            return response[0]["Name"]
```
Read my mind! I'll remove the `raise` and return `None`.
